spring-kafka/src/main/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainer.java

```diff
@@ -89,6 +95,13 @@ public void setConcurrency(int concurrency) {
 		return Collections.unmodifiableList(this.containers);
 	}
 
+	@Override
+	public Map<MetricName, ? extends Metric> metrics() {
+		return this.containers.stream()
+				.flatMap(c -> c.metrics().values().stream())
+				.collect(Collectors.toMap(Metric::metricName, Function.identity()));
```

Comment:
@tsachev ,

thank you for contribution. The looks clear and straightforward!
Although I wonder if there is going to be some collision with metrics overrides if we have concurrency > 1. I mean different target `Consumer`s may have the same `MetricName`, so the value from the first will be overridden by the second.
That might not be a case though if `   * @param group       logical group name of the metrics to which this metric belongs` is different for different consumers, but I somehow would like to see some samples how it is in the reality.

Thanks

Reply:
Perhaps we should change the signature to `public Map<Integer, Map<MetricName, ? extends Metric>> metrics()`, with the KMLC always returning a singleton `Map` and the concurrent container returning a map keyed by consumer number.

We need resolution on this by today or tomorrow since our release candidate is due Friday.

Reply:
Or, perhaps, `public Map<String, Map<MetricName, ? extends Metric>> metrics()` where the key is the listener thread name - it might help with correlation.

Reply:
Well, in terms of Apache Kafka we might need just use `clientId` on the matter.
But on the other hand I would like to debug this and see if all these `MetricName` s are not crossed and let us really to distinguish the `consumer` they belong to.

Reply:
So, debug shows this:
```
key = {MetricName@4991} "MetricName [name=outgoing-byte-rate, group=consumer-node-metrics, description=, tags={client-id=consumer-2, node-id=node-2147483647}]"
 name = "outgoing-byte-rate"
 group = "consumer-node-metrics"
 description = ""
 tags = {LinkedHashMap@5307}  size = 2
  0 = {LinkedHashMap$Entry@5324} "client-id" -> "consumer-2"
  1 = {LinkedHashMap$Entry@5325} "node-id" -> "node-2147483647"
 hash = -260015487
```
Pay attention to that ` "client-id" -> "consumer-2"` tag. Therefore I was right. Even in this flatten mode from the `ConcurrentMessageListenerContainer` we get unique metrics.

However I agree that it would be much convenient to group them per consumer. From here the structure like `Map<String, Map<MetricName, ? extends Metric>> metrics()` is fine with me. However I still vote for the `client-id` as a main key.

Reply:
My initial idea was to group the metrics too, but I could not decide for myself if it should be a `client-id` or something else.
Note that there is no Kafka API we can use to get the _actual_ consumer client id, because they may genereate it in the constructor if not supplied. I  need to rely on the `client-id` metric tag key, which I doubt they would change, anyway.

I will update the PR in few minutes.